src/Microsoft.ML.Core/Data/Schema.cs

```diff
@@ -0,0 +1,346 @@
+
+            /// <summary>
+            /// The type of the column.
+            /// </summary>
+            public ColumnType Type { get; }
+
+            /// <summary>
+            /// The metadata of the column.
+            /// </summary>
+            public MetadataRow Metadata { get; }
+
+            public Column(string name, ColumnType type, MetadataRow metadata)
+            {
+                Contracts.CheckNonEmpty(name, nameof(name));
+                Contracts.CheckValue(type, nameof(type));
+                Contracts.CheckValueOrNull(metadata);
+
+                Name = name;
+                Type = type;
+                Metadata = metadata;
+            }
+        }
+
+        /// <summary>
+        /// The metadata of one <see cref="Column"/>.
+        /// </summary>
+        public sealed class MetadataRow
+        {
+            private readonly (Column column, Delegate getter)[] _values;
+
+            /// <summary>
+            /// The schema of the metadata row. It is different from the schema that the column belongs to.
+            /// </summary>
+            public Schema Schema { get; }
+
+            /// <summary>
+            /// Create a metadata row by supplying the schema columns and the getter delegates for all the values.
+            /// </summary>
+            public MetadataRow(IEnumerable<(Column column, Delegate getter)> values)
+            {
+                Contracts.CheckValue(values, nameof(values));
+                // Check all getters.
+                foreach (var (column, getter) in values)
+                {
+                    Contracts.CheckValue(column, nameof(column));
+                    Contracts.CheckValue(getter, nameof(getter));
+                    Utils.MarshalActionInvoke(CheckGetter<int>, column.Type.RawType, getter);
+                }
+                _values = values.ToArray();
+                Schema = new Schema(_values.Select(x => x.column));
+            }
+
+            private void CheckGetter<TValue>(Delegate getter)
+            {
+                var typedGetter = getter as ValueGetter<TValue>;
+                if (typedGetter == null)
+                    throw Contracts.ExceptParam(nameof(getter), $"Getter of type '{typeof(TValue)}' expected, but {getter.GetType()} found");
+            }
+
+            /// <summary>
+            /// Get a getter delegate for one value of the metadata row.
+            /// </summary>
+            public ValueGetter<TValue> GetGetter<TValue>(int col)
+            {
+                Contracts.CheckParam(0 <= col && col < _values.Length, nameof(col));
+                var typedGetter = _values[col].getter as ValueGetter<TValue>;
+                return typedGetter;
+            }
+
+            /// <summary>
+            /// Get the value of the metadata, by column index.
+            /// </summary>
+            public void GetValue<TValue>(int col, ref TValue value) => GetGetter<TValue>(col)(ref value);
```

Comment:
>GetGetter<TValue>(col) [](start = 71, length = 22)

Would this throw if TValue is wrong?

Reply:
Yes

---
In reply to: [224846502](https://github.com/dotnet/machinelearning/pull/1167#discussion_r224846502) [](ancestors = 224846502)

Reply:
It should at least

---
In reply to: [224847616](https://github.com/dotnet/machinelearning/pull/1167#discussion_r224847616) [](ancestors = 224847616,224846502)